File: VideoClub/src/main/java/com/videoClub/service/FilmService.java
package com.videoClub.service;

import java.util.List;

import com.videoClub.dto.FilmDTO;
import com.videoClub.model.Film;
import com.videoClub.model.RegisteredUser;
import com.videoClub.model.User;
import com.videoClub.model.drl.FinalReport;
import com.videoClub.model.drl.RecommendedFilm;
import com.videoClub.model.enumeration.Genre;

public interface FilmService {

	public Film save(FilmDTO filmDTO);
	public Film getOne(Long id);
	public List<Film> getAll();
	public List<Film> getTopRated(int number);
	public List<Film> getMostPopular(int number);
	public List<Film> getByName(String filmName);
	public List<Film> getByGenre(Genre genre);
	public List<Film> getByArtist(Long artistId);
	public void delete(Long id);
	public Film rateFilm(Long filmId, Integer rate, RegisteredUser user);
	public List<Film> saveFilmToFavourites(Long filmId, User user);
	public FinalReport getRecommendedInfo(FilmDTO film);
	public List<Film> save(List<Film> films);
	public List<RecommendedFilm> getRecommendedFilmsByDefault(RegisteredUser user, int number);
	public List<RecommendedFilm> getRecommendedFilmsByArtis(RegisteredUser user, Long artistId, int number);
	public List<RecommendedFilm> getRecommendedFilmsByGenre(RegisteredUser user, Genre genre, int number);
	public List<RecommendedFilm> getRecommendedFilms(RegisteredUser user, List<Film> films, int number, boolean strip);
}

File: VideoClubFront/src/app/services/film.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {Observable} from 'rxjs';
import { map } from "rxjs/operators";
import { MessageDto } from '../dto/message.dto';
import { Film } from '../model/film.model';
import { FilmDto } from '../dto/film.dto';
import { RecommendedFilm } from '../model/recommended-film.model';
import { FinalReport } from '../dto/final-report';

@Injectable({providedIn: 'root'})
export class FilmService {

  constructor(private http: HttpClient) {
    
  }

    createFilm = (data: FilmDto): Observable<MessageDto> => {
        return this.http.post<MessageDto>("/api/film", data).pipe(
        map( (res: any) => {
            return res;
        })  );
    }

    updateFilm = (data: FilmDto): Observable<Film> => {
        return this.http.put<Film>("/api/film", data).pipe(
        map( (res: any) => {
            return res;
        })  );
    }

    rateFilm = (id: string, rate: number): Observable<Film> => {
        return this.http.get<Film>(`/api/film/rate/${id}/${rate}`).pipe(
        map( (res: any) => {
            return res;
        })  );
    }

    addToFavourites = (id: string): Observable<Film[]> => {
      return this.http.get<Film[]>(`/api/film/favourites/${id}`).pipe(
        map( (res: any) => {
            return res;
        })  );
    }

    checkIfFavourite = (id: string): Observable<boolean> => {
        return this.http.get<boolean>(`/api/film/favourites/exists/${id}`).pipe(
          map( (res: any) => {
              return res;
          })  );
      }

    filmsRecommended = (num: number): Observable<RecommendedFilm[]> => {
    return this.http.get<RecommendedFilm[]>(`/api/films/recommended/${num}`).pipe(
      map( (res: any) => {
          return res;
      })  );
    }

    filmReport = (film: FilmDto): Observable<FinalReport> => {
    return this.http.post<FinalReport>(`/api/film/recommended/info`, film).pipe(
        map( (res: any) => {
            return res;
        })  );
    }

    getTopRated = (num: number): Observable<Film[]> => {
        return this.http.get<Film[]>(`/api/films/top_rated/${num}`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }
    getMostPopular = (num: number): Observable<Film[]> => {
        return this.http.get<Film[]>(`/api/films/most_popular/${num}`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }

    getByGenre = (genre: string): Observable<Film[]> => {
        return this.http.get<Film[]>(`/api/films/search/genre/${genre}`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }

    getByName = (name: string): Observable<Film[]> => {
        return this.http.get<Film[]>(`/api/films/search/name/${name}`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }

    getFilmInfo = (id: number): Observable<Film> => {
        return this.http.get<Film>(`/api/film/${id}`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }

    getFilms = (): Observable<Film[]> => {
        return this.http.get<Film[]>(`/api/films`).pipe(
          map( (res: any) => {
              return res;
          })  );
    }
}